src/Data/Fin/Relation/Unary/Top.agda

```diff
@@ -0,0 +1,129 @@
+-- The View, considered as a unary relation on Fin (suc n)
+
+data View : (i : Fin (suc n)) → Set where
+
+  top :                View (fromℕ n)
+  inj : (j : Fin n) → View (inject₁ j)
+
+-- The view covering function, witnessing soundness of the view
+
+view : ∀ {n} i → View {n} i
+view {n = zero}  zero    = top
+view {n = suc _} zero    = inj _
+view {n = suc n} (suc i) with view {n} i
+... | top   = top
+... | inj j = inj (suc j)
+
+-- Interpretation of the view constructors
+
+⟦_⟧ : ∀ {i} → View {n} i → Fin (suc n)
+⟦ top ⟧   = fromℕ _
+⟦ inj j ⟧ = inject₁ j
+
+-- Completeness of the view
+
+view-complete : ∀ {i} (v : View {n} i) → ⟦ v ⟧ ≡ i
+view-complete top     = refl
+view-complete (inj j) = refl
+
+-- 'Computational' behaviour of the covering function
+
+view-top : ∀ n → view {n} (fromℕ n) ≡ top
+view-top zero    = refl
+view-top (suc n) rewrite view-top n = refl
+
+view-inj : ∀ j → view {n} (inject₁ j) ≡ inj j
+view-inj zero       = refl
+view-inj (suc j) rewrite view-inj j = refl
+
+------------------------------------------------------------------------
+-- Experimental
+--
+-- Because we work --without-K, Agda's unifier will complain about
+-- attempts to match `refl` against hypotheses of the form
+--    `view {n] i ≡ v`
+-- or gets stuck trying to solve unification problems of the form
+--    (inferred index ≟ expected index)
+-- even when these problems are *provably* solvable.
+--
+-- So the two predicates on values of the view defined below
+-- are extensionally equivalent to the assertions
+-- `view {n] i ≡ top` and `view {n] i ≡ inj j`
+--
+-- But such assertions can only ever have a unique (irrelevant) proof
+-- so we introduce instances to witness them, themselves given in
+-- terms of the functions `view-top` and `view-inj` defined above
+
+module Instances {n} where
+
+  data IsTop : ∀ {i} → View {n} i → Set where
+
+    top : IsTop top
+
+  instance
+
+    top⁺ : IsTop {i = fromℕ n} (view (fromℕ n))
+    top⁺ rewrite view-top n = top
+
+  data IsInj : ∀ {i} → View {n} i → Set where
+
+    inj : ∀ j → IsInj (inj j)
+
+  instance
+
+    inj⁺ : ∀ {j} → IsInj (view (inject₁ j))
+    inj⁺ {j} rewrite view-inj j = inj j
+
+    inject₁≡⁺ : ∀ {i} {j} {eq : inject₁ i ≡ j} → IsInj (view j)
+    inject₁≡⁺ {eq = refl} = inj⁺
+
+    inject₁≢n⁺ : ∀ {i} {n≢i : n ≢ toℕ (inject₁ i)} → IsInj (view {n} i)
+    inject₁≢n⁺ {i} {n≢i} with view i
+    ... | top   = ⊥-elim (n≢i (begin
+      n                         ≡˘⟨ toℕ-fromℕ n ⟩
+      toℕ (fromℕ n)           ≡˘⟨ toℕ-inject₁ (fromℕ n) ⟩
+      toℕ (inject₁ (fromℕ n)) ∎)) where open ≡-Reasoning
+    ... | inj j = inj j
+
+open Instances public
+
+------------------------------------------------------------------------
+-- As a corollary, we obtain two useful properties, which are
+-- witnessed by, but can also serve as proxy replacements for,
+-- the corresponding properties in `Data.Fin.Properties`
+
+module _ {n} where
```

Comment:
you have a `variable` declaration for `n` up top already, so it this `module _` needed?

Reply:
Good catch! For the cost of an additional `{n = n}` argument prefix in the first definition.